<!DOCTYPE html>
<html>
  <head>
    <title>Bio M</title>
    
      <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no">
<meta name="generator" content="Hugo 0.31.1" />

<title>使用BioPython将GenBank转换为Fasta格式 :: Bio M</title>
<link rel="shortcut icon" href="/images/favicon.png" type="image/x-icon" />
<link href="/css/nucleus.css" rel="stylesheet">
<link href="/css/font-awesome.min.css" rel="stylesheet">
<link href="/css/hybrid.css" rel="stylesheet">
<link href="/css/featherlight.min.css" rel="stylesheet">
<link href="/css/auto-complete.css" rel="stylesheet">
<link href="/theme-original/style.css" rel="stylesheet">

  <link href="/theme-original/variant-blue.css" rel="stylesheet">

<link rel="stylesheet" href="/css/bootstrap.min.css">
<script src="/js/jquery-2.x.min.js"></script>
<style type="text/css">
  :root #header + #content > #left > #rlblock_left {
    display:none !important;
  }
</style>
<meta name="description" content="">


    
  </head>
  <body data-url="/webpage_001/2014-09-28-transform-genbank-to-fasta-using-biopython/">
    
      <div id="headermain"></div>
<nav id="sidebar" class="showVisitedLinks">



<div class="highlightable">
  <div id="header-wrapper">
    <div id="header">
      
	
  
    <p><strong>Bio M</strong></p>

  

    </div>
        <div class="searchbox">
		    <label for="search-by"><i class="fa fa-search"></i></label>
		    <input data-search-input id="search-by" type="text" placeholder="Search...">
		    <span data-search-clear=""><i class="fa fa-close"></i></span>
		</div>
		<script type="text/javascript" src="/js/lunr.min.js"></script>
		<script type="text/javascript" src="/js/auto-complete.js"></script>
		<script type="text/javascript">
        
            var baseurl = "\/";
        
		</script>
		<script type="text/javascript" src="/js/search.js"></script>
  </div>

      <ul class="topics">
            <li data-nav-id="/" class="dd-item">
            <a href="/"><i class="fa fa-fw fa-home"></i></a>
            </li>
    <li data-nav-id="/biology/" class="dd-item">
      <a href="/biology/">
        <span>Biology</span><i class="fa fa-circle-thin read-icon"></i>
      </a>
    </li>
    <li data-nav-id="/linux/" class="dd-item">
      <a href="/linux/">
        <span>Linux</span><i class="fa fa-angle-right fa-lg category-icon"></i><i class="fa fa-circle-thin read-icon"></i>
      </a>
        <ul>
      <li data-nav-id="/linux/virtualbox_network/" class="dd-item">
        <a href="/linux/virtualbox_network/">
          <span>VirtualBox network</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/linux/git_learning/" class="dd-item">
        <a href="/linux/git_learning/">
          <span>Git examples</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/linux/git_usage/" class="dd-item">
        <a href="/linux/git_usage/">
          <span>Git usage</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/linux/ssh_connect_to_virtualbox/" class="dd-item">
        <a href="/linux/ssh_connect_to_virtualbox/">
          <span>SSH connect to VirtualBox</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/linux/ssh_connection_closed_by_remote/" class="dd-item">
        <a href="/linux/ssh_connection_closed_by_remote/">
          <span>SSH connection closed</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/linux/git_submodule/" class="dd-item">
        <a href="/linux/git_submodule/">
          <span>Git submodule</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
        </ul>
    </li>
    <li data-nav-id="/webpage_001/" class="dd-item parent">
      <a href="/webpage_001/">
        <span>webpage 001</span>
            <i class="fa fa-angle-down fa-lg category-icon"></i><i class="fa fa-circle-thin read-icon"></i>
      </a>
        <ul>
      <li data-nav-id="/webpage_001/2015-01-30-using-ensembl-genomes-perl-api/" class="dd-item">
        <a href="/webpage_001/2015-01-30-using-ensembl-genomes-perl-api/">
          <span>Using Ensembl Genomes Perl API</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-14-things-to-do-after-install-ubuntu-1404/" class="dd-item">
        <a href="/webpage_001/2014-10-14-things-to-do-after-install-ubuntu-1404/">
          <span>安装Ubuntu 14.04 之后需要做的几件事</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2015-03-26-small-rnas-in-bacteria/" class="dd-item">
        <a href="/webpage_001/2015-03-26-small-rnas-in-bacteria/">
          <span>EOB-001: Synthetic RNA silencing in bacteria</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2015-03-20-ggplot2parallel/" class="dd-item">
        <a href="/webpage_001/2015-03-20-ggplot2parallel/">
          <span>ggplot2 and parallel</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2015-04-13-manage-this-blog/" class="dd-item">
        <a href="/webpage_001/2015-04-13-manage-this-blog/">
          <span>Manage this blog</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2015-04-11-ssh-login-without-password/" class="dd-item">
        <a href="/webpage_001/2015-04-11-ssh-login-without-password/">
          <span>SSh login without password [2 steps]</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2015-03-29-toehold-switches/" class="dd-item">
        <a href="/webpage_001/2015-03-29-toehold-switches/">
          <span>EOB-002: Toehold switches De-Novo-Designed Regulators of Gene Expression</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2015-01-23-hypergeometric-analysis--enrichment-analysis/" class="dd-item">
        <a href="/webpage_001/2015-01-23-hypergeometric-analysis--enrichment-analysis/">
          <span>超几何分析和GO富集分析</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-11-14-install-rna-seq-analysis-tools/" class="dd-item">
        <a href="/webpage_001/2014-11-14-install-rna-seq-analysis-tools/">
          <span>RNA-Seq分析工具的配置</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-31-edit-linux-path/" class="dd-item">
        <a href="/webpage_001/2014-10-31-edit-linux-path/">
          <span>编辑Linux的$PATH</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-29-rnaseq-analysis-using-r-and-bioconductor/" class="dd-item">
        <a href="/webpage_001/2014-10-29-rnaseq-analysis-using-r-and-bioconductor/">
          <span>使用R和Bioconductor进行RNASeq差异表达分析</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-19-github-markdown-syntax/" class="dd-item">
        <a href="/webpage_001/2014-10-19-github-markdown-syntax/">
          <span>Github Markdown常用语法</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-19-add-search-box-to-blog/" class="dd-item">
        <a href="/webpage_001/2014-10-19-add-search-box-to-blog/">
          <span>在Jekyll blog中添加搜索框</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-08-open-source-mirrors-in-china/" class="dd-item">
        <a href="/webpage_001/2014-10-08-open-source-mirrors-in-china/">
          <span>国内可选择的开源软件镜像源</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-04-download-files-with-python/" class="dd-item">
        <a href="/webpage_001/2014-10-04-download-files-with-python/">
          <span>通过Python批量下载文件</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-04-download-files-with-r/" class="dd-item">
        <a href="/webpage_001/2014-10-04-download-files-with-r/">
          <span>通过R语言下载文档</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-28-transform-genbank-to-fasta-using-biopython/" class="dd-item active">
        <a href="/webpage_001/2014-09-28-transform-genbank-to-fasta-using-biopython/">
          <span>使用BioPython将GenBank转换为Fasta格式</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-23-mobaxterm-permissions-0660-for-id_rsa-are-too-open/" class="dd-item">
        <a href="/webpage_001/2014-09-23-mobaxterm-permissions-0660-for-id_rsa-are-too-open/">
          <span>MobaXterm: Permissions 0660 for &#39;id_rsa&#39; are too open</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-22-configuration-git-ssh/" class="dd-item">
        <a href="/webpage_001/2014-09-22-configuration-git-ssh/">
          <span>使用SSH管理Git</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-21-manage-multiple-account-on-github/" class="dd-item">
        <a href="/webpage_001/2014-09-21-manage-multiple-account-on-github/">
          <span>管理多个Github账号</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-20-jekyll&#43;github/" class="dd-item">
        <a href="/webpage_001/2014-09-20-jekyll&#43;github/">
          <span>jekyll&#43;Github</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-20-manage-github-pages/" class="dd-item">
        <a href="/webpage_001/2014-09-20-manage-github-pages/">
          <span>创建Github blogs</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-19-manage-blog-on-github/" class="dd-item">
        <a href="/webpage_001/2014-09-19-manage-blog-on-github/">
          <span>在Github上管理个人blog</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-10-r-read-compressed-files/" class="dd-item">
        <a href="/webpage_001/2014-09-10-r-read-compressed-files/">
          <span>R语言读取压缩文档</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-08-27-questions-about-using-rcurl/" class="dd-item">
        <a href="/webpage_001/2014-08-27-questions-about-using-rcurl/">
          <span>RCurl的问题</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
        </ul>
    </li>
    <li data-nav-id="/license/" class="dd-item">
      <a href="/license/">
        <span><i class='fa fa-copyright' aria-hidden='true'></i>License</span><i class="fa fa-circle-thin read-icon"></i>
      </a>
    </li>



        <section id="shortcuts">
                <li class="" role=""><h3>More</h3><a href="https://github.com/vjeantet/hugo-theme-docdock" target="_blank" rel="noopener"><i class='fa fa-github'></i> Github repo</a></li>
                <li class="" role=""><a href="http://genome.ucsc.edu/" target="_blank" rel="noopener"><i class='fa fa-bookmark'></i> UCSC Genome Browser</a></li>
        </section>
            <a id="clear-history" class="" href="#" data-clear-history-toggle=""><i class="fa  fa-history"></i> Clear History</a>

    <hr />
    <li></li>
    
    </ul>

 <section id="footer">
    </section>
  </div>
</nav>



<section id="body">
<div id="overlay"></div>
<div class="padding highlightable">

  <div id="top-bar">
    
      
      
      
    <div id="top-github-link">
      <a class="github-link" href="127.0.0.1webpage_001/2014-09-28-transform-genbank-to-fasta-using-biopython.md" target="blank">
        <i class="fa fa-code-fork"></i>
        Edit page
      </a>
    </div><div id="breadcrumbs" itemscope="" itemtype="http://data-vocabulary.org/Breadcrumb">
        <span id="sidebar-toggle-span">
          <a href="#" id="sidebar-toggle" data-sidebar-toggle="">
            <i class="fa fa-bars"></i>
          </a>
        </span>
        <span id="toc-menu"><i class="fa fa-list-alt"></i></span>
        <span class="links">
        







 <a href='/'>Bio M</a> > <a href='/webpage_001/'>webpage 001</a> > 使用BioPython将GenBank转换为Fasta格式

 

 

   
        </span>
    </div>
    
    
    <div class="progress">
        <div class="wrapper">
    <nav id="TableOfContents">
<ul>
<li><a href="#1-description">1 Description</a></li>
<li><a href="#2-purpose">2 Purpose</a></li>
<li><a href="#3-materials">3 Materials</a></li>
<li><a href="#4-methods-results">4 Methods &amp; Results</a></li>
</ul>
</nav>
        </div>
    </div>
    

  </div>


  <div id="tags">
    
      <a class="label label-default" href="/tags/biopython">Biopython</a>
    
  </div>

<div id="body-inner">
  
    <h1>使用BioPython将GenBank转换为Fasta格式</h1>
  



    
    
    
    

<p>Converting GenBank files to FASTA format with Biopython</p>

<h1 id="1-description">1 Description</h1>

<p>This post will show how to extrat info from GenBank file using Biopython, and convert to FASTA file (nucleotide sequences and amino acid sequence).</p>

<h1 id="2-purpose">2 Purpose</h1>

<ul>
<li>1. Extract entire DNA sequence as on record.<br /></li>
<li>2. Extract DNA sequences for each gene.<br /></li>
<li>3. Extract amino acid sequences for each gene.<br /></li>
</ul>

<h1 id="3-materials">3 Materials</h1>

<p>Here I choose a small bacteria complete genome to show the method: Buchnera_aphidicola (NCBI NC_011833, GI:219681389).
NCBI FTP <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/Bacteria/Buchnera_aphidicola_5A__Acyrthosiphon_pisum__uid59285">site </a></p>

<ul>
<li>GenBank file: <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/Bacteria/Buchnera_aphidicola_5A__Acyrthosiphon_pisum__uid59285/NC_011833.gbk">NC_011833.gbk</a> (2251 kb)<br /></li>
<li>Genome FASTA File: <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/Bacteria/Buchnera_aphidicola_5A__Acyrthosiphon_pisum__uid59285/NC_011833.fna">NC_011833.fna</a> (636 kb)<br /></li>
<li>Gene FASTA File: <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/Bacteria/Buchnera_aphidicola_5A__Acyrthosiphon_pisum__uid59285/NC_011833.ffn">NC_011833.ffn</a> (613 kb)<br /></li>
<li>Amino Acid File: <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/Bacteria/Buchnera_aphidicola_5A__Acyrthosiphon_pisum__uid59285/NC_011833.faa">NC_011833.faa</a> (246 kb)<br /></li>
</ul>

<h1 id="4-methods-results">4 Methods &amp; Results</h1>

<p>Using the module Bio.SeqIO to deal with genbank file.</p>

<ul>
<li>1. Genome: GenBank -&gt; FASTA<br />
Convert GenBank file to FASTA is simple：<br /></li>
</ul>

<p>Option 1:</p>

<pre><code class="language-python">&gt;&gt;&gt; from Bio import SeqIO  
&gt;&gt;&gt; SeqIO.convert(&quot;NC_011833.gbk&quot;, &quot;genbank&quot;, &quot;NC_011833_cov.fna&quot;, &quot;fasta&quot;)
</code></pre>

<p>Option 2:</p>

<pre><code class="language-python">import Bio.SeqIO
gbk_file = &quot;NC_011833.gbk&quot;
fna_file = &quot;NC_011833_cov.fna&quot;
input = open(gbk_file, &quot;r&quot;)
output = open(fna_file, &quot;w&quot;)

for seq in SeqIO.parse(input, &quot;genbank&quot;) :
    print &quot;Convert %s from GenBank to FASTA format&quot; % seq.id
	output.write(&quot;&gt;%s %s\n%s\n&quot; % (
        seq.id, seq.description, seq.seq))

input.close()
output.close()
</code></pre>

<p>This strain contains only one chromosome, so we can get one sequence using above code.</p>

<ul>
<li>2. Protein AA: GenBank -&gt; FASTA
Extract the sequences of each coding sequences and convert to FASTA format.<br />
We can use <strong>features</strong> to extract different information in GenBank file.</li>
</ul>

<pre><code class="language-python">import Bio.SeqIO

gbk_file = &quot;NC_011833.gbk&quot;
fna_file = &quot;NC_011833_cov.fna&quot;
faa_file = &quot;NC_011833_cov.faa&quot;
input = open(gbk_file, &quot;r&quot;)
geneNC = open(fna_file, &quot;w&quot;)
geneAA = open(faa_file, &quot;w&quot;)

for seq in SeqIO.parse(input, &quot;genbank&quot;) :
    print &quot;Dealing with GenBank file of %s, \nOutput: \ngene nc: %s \ngene aa: %s&quot; % (
        seq.id,
        fna_file,
        faa_file)
    for seq_feature in seq.features :
        geneSeq = seq_feature.extract(seq.seq)
        if seq_feature.type == &quot;CDS&quot; :
            assert len(seq_feature.qualifiers['translation']) == 1
            geneAA.write(&quot;&gt;%s %s, %s\n%s\n&quot; % (
                seq_feature.qualifiers['locus_tag'][0],
                seq.name,
                seq.description,
                seq_feature.qualifiers['translation'][0]))
            geneNC.write(&quot;&gt;%s %s, %s\n%s\n&quot; % (
                seq_feature.qualifiers['locus_tag'][0],
                seq.name,
                seq.description,
                geneSeq ))

input.close()
geneNC.close()
geneAA.close()
</code></pre>

<p>The result will contain FASTA record of each CDSin each GenBank record.<br />
Note:
Not all CDS features of GenBank file contain translation.  I think we can extract the sequences of each gene, and translate them to amino acid sequences.</p>

<ul>
<li>3. Gene : GenBank -&gt; FASTA
We will use seq.extract() to get sequence.</li>
</ul>

<pre><code class="language-python">import Bio.SeqIO

gbk_file = &quot;NC_011833.gbk&quot;
fna_file = &quot;NC_011833_cov.fna&quot;
faa_file = &quot;NC_011833_cov.faa&quot;
input = open(gbk_file, &quot;r&quot;)
geneNC = open(fna_file, &quot;w&quot;)
geneAA = open(faa_file, &quot;w&quot;)

for seq in SeqIO.parse(input, &quot;genbank&quot;) :
    print &quot;Dealing with GenBank file of %s, \nOutput: \ngene nc: %s \ngene aa: %s&quot; % (
        seq.id,
        fna_file,
        faa_file)
    for seq_feature in seq.features :
        geneSeq = seq_feature.extract(seq.seq)
        if seq_feature.type == &quot;CDS&quot; :
            assert len(seq_feature.qualifiers['translation']) == 1
            geneNC.write(&quot;&gt;%s %s, %s\n%s\n&quot; % (
                seq_feature.qualifiers['locus_tag'][0],
                seq.name,
                seq.description,
                geneSeq ))

input.close()
geneNC.close()
geneAA.close()
</code></pre>

<ul>
<li>Expand<br /></li>
</ul>

<p>Learn Biopython Bio.SeqIO usage:</p>

<ul>
<li>1. Peter cock, <a href="http://www2.warwick.ac.uk/fac/sci/moac/people/students/peter_cock/python/genbank2fasta/">http://www2.warwick.ac.uk/fac/sci/moac/people/students/peter_cock/python/genbank2fasta/</a><br /></li>
<li>2. Biopython cookbook, <a href="http://biopython.org/DIST/docs/tutorial/Tutorial.html">http://biopython.org/DIST/docs/tutorial/Tutorial.html</a><br /></li>
</ul>


    
    
          <footer class=" footline" >
	
	    <i class='fa fa-user'></i> <a href="mailto:wangm08@hotmail.com">Ming Wang</a>  <i class='fa fa-calendar'></i> 28/09/2014
	    </div>
	
</footer>
  </div>
</div>

<div id="navigation">
</div>

</section>
<div style="left: -1000px; overflow: scroll; position: absolute; top: -1000px; border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;">
  <div style="border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;"></div>
</div>    

<script src="/js/clipboard.min.js"></script>
<script src="/js/featherlight.min.js"></script>
<script src="/js/html5shiv-printshiv.min.js"></script>
<script src="/js/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
<script src="/js/modernizr.custom.71422.js"></script>
<script src="/js/docdock.js"></script>
<script src="/theme-original/script.js"></script>


    

    
    

    
  </body>
</html>