<!DOCTYPE html>
<html>
  <head>
    <title>Bio M</title>
    
      <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no">
<meta name="generator" content="Hugo 0.31.1" />

<title>使用R和Bioconductor进行RNASeq差异表达分析 :: Bio M</title>
<link rel="shortcut icon" href="/images/favicon.png" type="image/x-icon" />
<link href="/css/nucleus.css" rel="stylesheet">
<link href="/css/font-awesome.min.css" rel="stylesheet">
<link href="/css/hybrid.css" rel="stylesheet">
<link href="/css/featherlight.min.css" rel="stylesheet">
<link href="/css/auto-complete.css" rel="stylesheet">
<link href="/theme-original/style.css" rel="stylesheet">

  <link href="/theme-original/variant-blue.css" rel="stylesheet">

<link rel="stylesheet" href="/css/bootstrap.min.css">
<script src="/js/jquery-2.x.min.js"></script>
<style type="text/css">
  :root #header + #content > #left > #rlblock_left {
    display:none !important;
  }
</style>
<meta name="description" content="">


    
  </head>
  <body data-url="/webpage_001/2014-10-29-rnaseq-analysis-using-r-and-bioconductor/">
    
      <div id="headermain"></div>
<nav id="sidebar" class="showVisitedLinks">



<div class="highlightable">
  <div id="header-wrapper">
    <div id="header">
      
	
  
    <p><strong>Bio M</strong></p>

  

    </div>
        <div class="searchbox">
		    <label for="search-by"><i class="fa fa-search"></i></label>
		    <input data-search-input id="search-by" type="text" placeholder="Search...">
		    <span data-search-clear=""><i class="fa fa-close"></i></span>
		</div>
		<script type="text/javascript" src="/js/lunr.min.js"></script>
		<script type="text/javascript" src="/js/auto-complete.js"></script>
		<script type="text/javascript">
        
            var baseurl = "\/";
        
		</script>
		<script type="text/javascript" src="/js/search.js"></script>
  </div>

      <ul class="topics">
            <li data-nav-id="/" class="dd-item">
            <a href="/"><i class="fa fa-fw fa-home"></i></a>
            </li>
    <li data-nav-id="/biology/" class="dd-item">
      <a href="/biology/">
        <span>Biology</span><i class="fa fa-circle-thin read-icon"></i>
      </a>
    </li>
    <li data-nav-id="/linux/" class="dd-item">
      <a href="/linux/">
        <span>Linux</span><i class="fa fa-angle-right fa-lg category-icon"></i><i class="fa fa-circle-thin read-icon"></i>
      </a>
        <ul>
      <li data-nav-id="/linux/virtualbox_network/" class="dd-item">
        <a href="/linux/virtualbox_network/">
          <span>VirtualBox network</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/linux/git_learning/" class="dd-item">
        <a href="/linux/git_learning/">
          <span>Git examples</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/linux/git_usage/" class="dd-item">
        <a href="/linux/git_usage/">
          <span>Git usage</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/linux/ssh_connect_to_virtualbox/" class="dd-item">
        <a href="/linux/ssh_connect_to_virtualbox/">
          <span>SSH connect to VirtualBox</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/linux/ssh_connection_closed_by_remote/" class="dd-item">
        <a href="/linux/ssh_connection_closed_by_remote/">
          <span>SSH connection closed</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/linux/git_submodule/" class="dd-item">
        <a href="/linux/git_submodule/">
          <span>Git submodule</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
        </ul>
    </li>
    <li data-nav-id="/webpage_001/" class="dd-item parent">
      <a href="/webpage_001/">
        <span>webpage 001</span>
            <i class="fa fa-angle-down fa-lg category-icon"></i><i class="fa fa-circle-thin read-icon"></i>
      </a>
        <ul>
      <li data-nav-id="/webpage_001/2015-01-30-using-ensembl-genomes-perl-api/" class="dd-item">
        <a href="/webpage_001/2015-01-30-using-ensembl-genomes-perl-api/">
          <span>Using Ensembl Genomes Perl API</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-14-things-to-do-after-install-ubuntu-1404/" class="dd-item">
        <a href="/webpage_001/2014-10-14-things-to-do-after-install-ubuntu-1404/">
          <span>安装Ubuntu 14.04 之后需要做的几件事</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2015-03-26-small-rnas-in-bacteria/" class="dd-item">
        <a href="/webpage_001/2015-03-26-small-rnas-in-bacteria/">
          <span>EOB-001: Synthetic RNA silencing in bacteria</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2015-03-20-ggplot2parallel/" class="dd-item">
        <a href="/webpage_001/2015-03-20-ggplot2parallel/">
          <span>ggplot2 and parallel</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2015-04-13-manage-this-blog/" class="dd-item">
        <a href="/webpage_001/2015-04-13-manage-this-blog/">
          <span>Manage this blog</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2015-04-11-ssh-login-without-password/" class="dd-item">
        <a href="/webpage_001/2015-04-11-ssh-login-without-password/">
          <span>SSh login without password [2 steps]</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2015-03-29-toehold-switches/" class="dd-item">
        <a href="/webpage_001/2015-03-29-toehold-switches/">
          <span>EOB-002: Toehold switches De-Novo-Designed Regulators of Gene Expression</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2015-01-23-hypergeometric-analysis--enrichment-analysis/" class="dd-item">
        <a href="/webpage_001/2015-01-23-hypergeometric-analysis--enrichment-analysis/">
          <span>超几何分析和GO富集分析</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-11-14-install-rna-seq-analysis-tools/" class="dd-item">
        <a href="/webpage_001/2014-11-14-install-rna-seq-analysis-tools/">
          <span>RNA-Seq分析工具的配置</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-31-edit-linux-path/" class="dd-item">
        <a href="/webpage_001/2014-10-31-edit-linux-path/">
          <span>编辑Linux的$PATH</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-29-rnaseq-analysis-using-r-and-bioconductor/" class="dd-item active">
        <a href="/webpage_001/2014-10-29-rnaseq-analysis-using-r-and-bioconductor/">
          <span>使用R和Bioconductor进行RNASeq差异表达分析</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-19-github-markdown-syntax/" class="dd-item">
        <a href="/webpage_001/2014-10-19-github-markdown-syntax/">
          <span>Github Markdown常用语法</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-19-add-search-box-to-blog/" class="dd-item">
        <a href="/webpage_001/2014-10-19-add-search-box-to-blog/">
          <span>在Jekyll blog中添加搜索框</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-08-open-source-mirrors-in-china/" class="dd-item">
        <a href="/webpage_001/2014-10-08-open-source-mirrors-in-china/">
          <span>国内可选择的开源软件镜像源</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-04-download-files-with-python/" class="dd-item">
        <a href="/webpage_001/2014-10-04-download-files-with-python/">
          <span>通过Python批量下载文件</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-10-04-download-files-with-r/" class="dd-item">
        <a href="/webpage_001/2014-10-04-download-files-with-r/">
          <span>通过R语言下载文档</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-28-transform-genbank-to-fasta-using-biopython/" class="dd-item">
        <a href="/webpage_001/2014-09-28-transform-genbank-to-fasta-using-biopython/">
          <span>使用BioPython将GenBank转换为Fasta格式</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-23-mobaxterm-permissions-0660-for-id_rsa-are-too-open/" class="dd-item">
        <a href="/webpage_001/2014-09-23-mobaxterm-permissions-0660-for-id_rsa-are-too-open/">
          <span>MobaXterm: Permissions 0660 for &#39;id_rsa&#39; are too open</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-22-configuration-git-ssh/" class="dd-item">
        <a href="/webpage_001/2014-09-22-configuration-git-ssh/">
          <span>使用SSH管理Git</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-21-manage-multiple-account-on-github/" class="dd-item">
        <a href="/webpage_001/2014-09-21-manage-multiple-account-on-github/">
          <span>管理多个Github账号</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-20-jekyll&#43;github/" class="dd-item">
        <a href="/webpage_001/2014-09-20-jekyll&#43;github/">
          <span>jekyll&#43;Github</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-20-manage-github-pages/" class="dd-item">
        <a href="/webpage_001/2014-09-20-manage-github-pages/">
          <span>创建Github blogs</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-19-manage-blog-on-github/" class="dd-item">
        <a href="/webpage_001/2014-09-19-manage-blog-on-github/">
          <span>在Github上管理个人blog</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-09-10-r-read-compressed-files/" class="dd-item">
        <a href="/webpage_001/2014-09-10-r-read-compressed-files/">
          <span>R语言读取压缩文档</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
      <li data-nav-id="/webpage_001/2014-08-27-questions-about-using-rcurl/" class="dd-item">
        <a href="/webpage_001/2014-08-27-questions-about-using-rcurl/">
          <span>RCurl的问题</span><i class="fa fa-circle-thin read-icon"></i>
        </a>
    </li>
        </ul>
    </li>
    <li data-nav-id="/license/" class="dd-item">
      <a href="/license/">
        <span><i class='fa fa-copyright' aria-hidden='true'></i>License</span><i class="fa fa-circle-thin read-icon"></i>
      </a>
    </li>



        <section id="shortcuts">
                <li class="" role=""><h3>More</h3><a href="https://github.com/vjeantet/hugo-theme-docdock" target="_blank" rel="noopener"><i class='fa fa-github'></i> Github repo</a></li>
                <li class="" role=""><a href="http://genome.ucsc.edu/" target="_blank" rel="noopener"><i class='fa fa-bookmark'></i> UCSC Genome Browser</a></li>
        </section>
            <a id="clear-history" class="" href="#" data-clear-history-toggle=""><i class="fa  fa-history"></i> Clear History</a>

    <hr />
    <li></li>
    
    </ul>

 <section id="footer">
    </section>
  </div>
</nav>



<section id="body">
<div id="overlay"></div>
<div class="padding highlightable">

  <div id="top-bar">
    
      
      
      
    <div id="top-github-link">
      <a class="github-link" href="127.0.0.1webpage_001/2014-10-29-rnaseq-analysis-using-r-and-bioconductor.md" target="blank">
        <i class="fa fa-code-fork"></i>
        Edit page
      </a>
    </div><div id="breadcrumbs" itemscope="" itemtype="http://data-vocabulary.org/Breadcrumb">
        <span id="sidebar-toggle-span">
          <a href="#" id="sidebar-toggle" data-sidebar-toggle="">
            <i class="fa fa-bars"></i>
          </a>
        </span>
        <span id="toc-menu"><i class="fa fa-list-alt"></i></span>
        <span class="links">
        







 <a href='/'>Bio M</a> > <a href='/webpage_001/'>webpage 001</a> > 使用R和Bioconductor进行RNASeq差异表达分析

 

 

   
        </span>
    </div>
    
    
    <div class="progress">
        <div class="wrapper">
    <nav id="TableOfContents">
<ul>
<li><a href="#count-based-differential-expression-analysis-of-rna-sequencing-data-using-r-and-bioconductor">Count-based differential expression analysis of RNA sequencing data using R and Bioconductor</a></li>
<li><a href="#1-summary">1 Summary</a></li>
<li><a href="#2-method">2 Method</a></li>
</ul>
</nav>
        </div>
    </div>
    

  </div>


  <div id="tags">
    
      <a class="label label-default" href="/tags/r">R</a>
    
      <a class="label label-default" href="/tags/rnaseq">RNASeq</a>
    
  </div>

<div id="body-inner">
  
    <h1>使用R和Bioconductor进行RNASeq差异表达分析</h1>
  



    
    
    
    

<p>RNA-Seq analysis using R and Bioconductor</p>

<h1 id="count-based-differential-expression-analysis-of-rna-sequencing-data-using-r-and-bioconductor">Count-based differential expression analysis of RNA sequencing data using R and Bioconductor</h1>

<p>Published online 22 August 2013; doi:10.1038/nprot.2013.099
nature protocols | VOL.8 NO.9 | 2013</p>

<p>See the original material: <a href="http://www.nature.com/nprot/journal/v8/n9/full/nprot.2013.099.html">URL</a></p>

<h1 id="1-summary">1 Summary</h1>

<p>The samples cited here can be found @ <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3032923/">URL</a><br />
<em>Brooks, A.N. et al. Conservation of an RNA regulatory map between Drosophila and mammals. Genome Res. 21, 193–202 (2011).</em></p>

<p>Library           Library Layout    shortname     Fastaq</p>

<hr />

<p>Untreated-3       PAIRED            CT.PA.1       SRR031714, *15<br />
Untreated-4       PAIRED            CT.PA.2       SRR031716, *17<br />
CG8144_RNAi-3     PAIRED            KD.PA.3       SRR031724, *25<br />
CG8144_RNAi-4     PAIRED            KD.PA.4       SRR031726, *27<br />
Untreated-1       SINGLE            CT.SI.5       SRR031708, *09-13<br />
CG8144_RNAi-1     SINGLE            KD.SI.6       SRR031718, *19-23<br />
Untreated-6       SINGLE            CT.SI.7       SRR031728, *29</p>

<hr />

<h1 id="2-method">2 Method</h1>

<p>*1 Go to work directory</p>

<pre><code class="language-R">setwd(&quot;/home/wangming/work/R_bioconductor/&quot;)
</code></pre>

<p>*2 Loading sample information</p>

<pre><code class="language-R">readRDS(&quot;sri.txt&quot;)
</code></pre>

<p><strong>Start procedure</strong></p>

<p>*3 Reads quality control</p>

<pre><code class="language-R">#library(&quot;ShortRead&quot;)
#fqQC &lt;- qa(dirPath=&quot;reads&quot;, pattern=&quot;.fastq$&quot;, type=&quot;fastq&quot;)
#report(fqQC, type=&quot;html&quot;, dest=&quot;fastqQAreport&quot;)
</code></pre>

<p>*4 Create sample table</p>

<pre><code class="language-R">sri$LibraryName &lt;- gsub(&quot;S2_DRSC_&quot;, &quot;&quot;, sri$LibraryName) # trim the library name
samples &lt;- unique(sri[, c(&quot;LibraryName&quot;, &quot;LibraryLayout&quot;)])
for(i in 1:nrow(samples)) {
  rw &lt;- (sri$LibraryName == samples$LibraryName[i])
  if(samples$LibraryLayout[i] == &quot;PAIRED&quot;) {
    samples$fastq1[i]=paste0(paste(&quot;reads&quot;, sri$Run[rw], sep=&quot;/&quot;), &quot;_1.fastq&quot;, collapse=&quot;,&quot;)
    samples$fastq2[i]=paste0(paste(&quot;reads&quot;, sri$Run[rw], sep=&quot;/&quot;), &quot;_2.fastq&quot;, collapse=&quot;,&quot;)    
  } else {
    samples$fastq1[i]=paste0(paste(&quot;reads&quot;, sri$Run[rw], sep=&quot;/&quot;), &quot;.fastq&quot;, collapse=&quot;,&quot;)
    samples$fastq2[i]=&quot;&quot;
  }
}
</code></pre>

<p>*5 Add description columns</p>

<pre><code class="language-R">samples$condition &lt;- &quot;CTL&quot;
samples$condition[grep(&quot;RNAi&quot;, samples$LibraryName)] &lt;- &quot;KD&quot; # Knock down
samples$shortname &lt;- paste(substr(samples$condition, 1, 2), 
                           substr(samples$LibraryLayout, 1, 2),
                           seq_len(nrow(samples)), sep=&quot;.&quot;)

</code></pre>

<p>*6 Inspect the table carefully<br />
<em>print out the &ldquo;samples&rdquo;</em></p>

<pre><code class="language-R">head(samples, 5)
</code></pre>

<p>*7 Using R to create shell commands</p>

<pre><code class="language-R">gtf &lt;- &quot;reference/Drosophila_melanogaster.BDGP5.70.gtf&quot;
bowtieIndex &lt;- &quot;reference/Dme1_BDGP5_70&quot;
bowtieCMD &lt;- with(samples, paste(&quot;tophat2 -G&quot;, gtf, &quot;-p 5 -o&quot;, 
                                 LibraryName, bowtieIndex, fastq1, fastq2))
bowtieCMD
</code></pre>

<p>*8 Run the command in Unix terminal</p>

<pre><code class="language-R">write.table(bowtieCMD, &quot;RunBowtie2.R&quot;, sep=&quot;\n&quot;, row.names=FALSE, col.names=FALSE)
</code></pre>

<p>*9 Organize the BAM files and create SAM files</p>

<pre><code class="language-R">smCMD &lt;- &quot;&quot;
for(i in 1:nrow(samples)) {
  lib &lt;- samples$LibraryName[i]
  ob &lt;- file.path(lib, &quot;accepted_hits.bam&quot;)
  # sort bam files by name, convert to SAM for htseq-count
  sm1 &lt;- paste0(&quot;samtools sort -n &quot;, ob, &quot; &quot;, lib, &quot;_sn&quot;)
  sm2 &lt;- paste0(&quot;samtools view -o &quot;, lib, &quot;_sn.sam &quot;, lib, &quot;_sn.bam&quot;)
  # sort by position and index for IGV
  sm3 &lt;- paste0(&quot;samtools sort &quot;, ob, &quot; &quot;, lib, &quot;_s&quot;)
  sm4 &lt;- paste0(&quot;samtools index &quot;, lib, &quot;_s.bam&quot;)
  smCMD[i] &lt;- paste(sm1, sm2, sm3, sm4, sep=&quot;\n&quot;)
  write.table(smCMD[i], paste0(&quot;RunSm-&quot;, samples$LibraryName[i], &quot;.sh&quot;), 
              row.names=FALSE, col.names=FALSE)
}
#write.table(smCMD, &quot;RunSamtools.sh&quot;, sep=&quot;\n\n&quot;, row.names=FALSE, col.names=FALSE)

</code></pre>

<p>Run the above commands in UNIX terminal.</p>

<p>*10 Inspect the Alignment data using IGV
<em>D. melanogaster (dm3) genome, BAM files and GTF file.</em></p>

<p>*11 Check the expressed transcript (Figure 3)</p>

<p>*12 Select the positive and negative controls for the system from other study.</p>

<p>*13 Count reads using htseq-count</p>

<pre><code class="language-R">samples$countf &lt;- paste(samples$LibraryName, &quot;count&quot;, sep=&quot;.&quot;)
gtf &lt;- &quot;reference/Drosophila_melanogaster.BDGP5.70.gtf&quot;
htseqCMD &lt;- paste0(&quot;htseq-count -s no -a 10 &quot;, samples$LibraryName, 
                   &quot;_sn.sam &quot;, gtf, &quot; &gt; &quot;, samples$countf)
write.table(htseqCMD, &quot;HTSeq-count.sh&quot;, row.names=FALSE, col.names=FALSE)
htseqCMD
</code></pre>

<p>Run the above commands in UNIX terminal.</p>

<p>*14 Differential expression analysis using edgeR.
A. edgeR - simple
B. edgeR - complex
C. DESeq - simple
D. DESeq - complex</p>

<p>*A. edgeR - simple</p>

<pre><code class="language-R">##1 Start edgeR
library(&quot;edgeR&quot;)
counts &lt;- readDGE(samples$countf)$counts
##2 Filter weakly expressed and noninformative features
noint &lt;- rownames(counts) %in% c(&quot;no_feature&quot;, &quot;ambiguous&quot;, &quot;too_low_aQual&quot;,
                                 &quot;alignment_not_unique&quot;)
cpms &lt;- cpm(counts)
keep &lt;- rowSums(cpms &gt; 1) &gt;= 3 &amp; ! noint
counts &lt;- counts[keep, ]
##2 Inspect the count table
colnames(counts) &lt;- samples$shortname
head(counts[, order(samples$condition)], 5)
##4 Create a DGEList object.
d &lt;- DGEList(counts = counts, group = samples$condition)
##5 Estimate normalization factors
d &lt;- calcNormFactors(d)
##6 Inspect the relationships using multidimentional scaling (MDS) plot
plotMDS(d, labels = samples$shortname,
        col = c(&quot;darkgreen&quot;,&quot;blue&quot;)[factor(samples$condition)])
##7 Estimate tagwise dispersion (simple design)
d &lt;- estimateCommonDisp(d)
d &lt;- estimateTagwiseDisp(d)
##8 Create a visual representation of the mean-variance relationship using the 
##  plotMeanVar (Fig. 5a) and plotBCV (Fig. 5b) functions, as follows:
plotMeanVar(d, show.tagwise.vars = TRUE, NBline = TRUE)
plotBCV(d)
##9 Test for differential expression (‘classic’ edgeR),
de &lt;- exactTest(d, pair = c(&quot;CTL&quot;,&quot;KD&quot;))
##10 See Step 14B(##6-##9)
##//---Step 14B(##6-##9)---##
##6 Use the topTags function to present a tabular summary of the differential 
##  expression statistics.
tt &lt;- topTags(de, n=nrow(d))
head(tt$table)
##7 Inspect the depth-adjusted reads per million for some of the top 
##  differentially expressed genes
nc &lt;- cpm(d, normalized.lib.sizes=TRUE)
rn &lt;- rownames(tt$table)
head(nc[rn, order(samples$condition)], 5)
##8 Create a graphical summary, such as an M (log-fold) VS A (log-average) plot,
##  here showing the genes selected as differentially expressed (FDR=0.05)
deg &lt;- rn[tt$table$FDR &lt; 0.05]
plotSmear(d, de.tags=deg)
##9 Save the result table as CSV file
write.csv(tt$table, file=&quot;toptags_edgeR.csv&quot;)
##//------##
</code></pre>

<p>*B. edgeR - complex</p>

<pre><code class="language-R">##1 Follow Step 14A(##1-##6)
##2 Create a design matrix (see ‘Experimental design’ for further details) to 
## specify the factors that are expected to affect expression levels:
design = model.matrix( ~ LibraryLayout + condition, samples)
design
##3 Estimate dispersion values, using the Cox-Reid (CR)-adjusted likelihood
d2 &lt;- estimateGLMTrendedDisp(d, design)
d3 &lt;- estimateGLMTagwiseDisp(d2, design)
##4 Given the design matrix and dispersion esitmates, fit a GLM to each feature
f &lt;- glmFit(d2, design)
##5 Perform a likelihood ratio test, specifying the difference of interest
##  (Knockdown VS control, the 3rd column of the matrix)
de &lt;- glmLRT(f, coef=3)
##6 Use the topTags function to present a tabular summary of the differential 
##  expression statistics.
tt &lt;- topTags(de, n=nrow(d))
head(tt$table)
##7 Inspect the depth-adjusted reads per million for some of the top 
##  differentially expressed genes
nc &lt;- cpm(d, normalized.lib.sizes=TRUE)
rn &lt;- rownames(tt$table)
head(nc[rn, order(samples$condition)], 5)
##8 Create a graphical summary, such as an M (log-fold) VS A (log-average) plot,
##  here showing the genes selected as differentially expressed (FDR=0.05)
deg &lt;- rn[tt$table$FDR &lt; 0.05]
plotSmear(d, de.tags=deg)
##9 Save the result table as CSV file
write.csv(tt$table, file=&quot;toptags_edgeR.csv&quot;)
</code></pre>

<p>*C. DESeq - simple design</p>

<pre><code class="language-R">##1 Create a data.frame with required data: 
samplesDESeq &lt;- with(samples, 
                     data.frame(shortname=I(shortname), countf=I(countf),
                                condition=condition,
                                LibraryLayout=LibraryLayout))
##2 Load the DESeq package and create a CountDataSet object (DESeq’s container 
##  for RNA-seq data) from the count tables and corresponding metadata:
library(&quot;DESeq&quot;)
cds &lt;- newCountDataSetFromHTSeqCount(samplesDESeq)
##3 Estimate normalization factors
cds &lt;- estimateSizeFactors(cds)
##4 Inspect the size factors
sizeFactors(cds)
##5 Inspect sample relationships, invoke a variance-stabilizing trasformation
##  and inspect a principal component analysis (PCA) plot
cdsB &lt;- estimateDispersions(cds, method=&quot;blind&quot;)
vsd &lt;- varianceStabilizingTransformation(cdsB)
p &lt;- plotPCA(vsd, intgroup=c(&quot;condition&quot;, &quot;LibraryLayout&quot;))
##6 Use estimateDispersions to calculate dispersion values
cds &lt;- estimateDispersions(cds)
##7 Inspect the estimated dispersions using the plotDispEsts function
plotDispEsts(cds)
##8 Perform the test for differential expression by using nbinomTest
res &lt;- nbinomTest(cds,&quot;CTL&quot;,&quot;KD&quot;)
##9 Given the table of differential expression results, use plotMA to display
##  differential expression (log-fold) VS expression strength (log-average)
plotMA(res)
##10 Inspect the result tables of significantly upregulated and downregulated 
##   genes: FDR=0.1
resSig &lt;- res[which(res$padj &lt; 0.1), ]
head(resSig[order(resSig$log2FoldChange, decreasing=TRUE), ])
head(resSig[order(resSig$log2FoldChange, decreasing=FALSE), ])
##11 Count the number of genes with significant differential expression at 
##   FDR=0.1
table(res$padj &lt; 0.1)
##12 Create a persistent storage of results using a CSV file
write.csv(res, file=&quot;res_DESeq.csv&quot;)
##13 Perform a sanity check by inspecting a histogram of unadjusted P values
##   for the differential expression results
hist(res$pval, breaks=100)
</code></pre>

<p>*D. DESeq - complex design</p>

<pre><code class="language-R">##1 Follow Step 14C(##1-##5)
##2 Calculate the CR-adjusted profile likelihood dispersion estimates relative 
##  to the factors specified, developed by McCarthy et al., according to:
cds &lt;- estimateDispersions(cds, method = &quot;pooled-CR&quot;,
                              modelFormula = count ~ LibraryLayout + condition)
##3 Test for differential expression in the GLM setting by fitting both a full 
##  model and reduced model (i.e., with the factor of interest taken out):
fit1 &lt;- fitNbinomGLMs(cds, count ~ LibraryLayout + condition)
fit0 &lt;- fitNbinomGLMs(cds, count ~ LibraryLayout)
##4 By using the two fitted models, compute likelihood ratio statistics and 
##  associated P values, as follows:
pval &lt;- nbinomGLMTest(fit1, fit0)
##5 Adjust the reported P values for multiple testing:
padj &lt;- p.adjust(pval, method = &quot;BH&quot;)
##6 Assemble a result table from full model fit and the raw and adjusted P 
##  values and print the first few upregulated and downregulated genes (FDR &lt; 10%):
res &lt;- cbind(fit1, pval=pval, padj=padj)
resSig &lt;- res[which(res$padj &lt; 0.1),]
head(resSig[order(resSig$conditionKD, decreasing=TRUE), ])
head(resSig[order(resSig$conditionKD, decreasing=FALSE), ])
##7 Follow Step 14C(##11-##13).
##//------##
##11 Count the number of genes with significant differential expression at 
##   FDR=0.1
table(res$padj &lt; 0.1)
##12 Create a persistent storage of results using a CSV file
write.csv(res, file=&quot;res_DESeq.csv&quot;)
##13 Perform a sanity check by inspecting a histogram of unadjusted P values
##   for the differential expression results
hist(res$pval, breaks=100)
##//------##
</code></pre>

<p>*15. As another spot check, point the IGV genome browser (with GTF and BAM
files loaded) to a handful of the top differentially expressed genes and
confirm that the counting and differential expression statistics are
appropriately represented.</p>


    
    
          <footer class=" footline" >
	
	    <i class='fa fa-user'></i> <a href="mailto:wangm08@hotmail.com">Ming Wang</a>  <i class='fa fa-calendar'></i> 29/10/2014
	    </div>
	
</footer>
  </div>
</div>

<div id="navigation">
</div>

</section>
<div style="left: -1000px; overflow: scroll; position: absolute; top: -1000px; border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;">
  <div style="border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;"></div>
</div>    

<script src="/js/clipboard.min.js"></script>
<script src="/js/featherlight.min.js"></script>
<script src="/js/html5shiv-printshiv.min.js"></script>
<script src="/js/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
<script src="/js/modernizr.custom.71422.js"></script>
<script src="/js/docdock.js"></script>
<script src="/theme-original/script.js"></script>


    

    
    

    
  </body>
</html>